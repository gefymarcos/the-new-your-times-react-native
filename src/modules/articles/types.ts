export type ArticleProps = {
  title: string,
  abstract: string,
  url: string
}

export type ArticlesStoreProps = {
  articlesRequesting: boolean,
  errored: boolean,
  errorMsg: string,
  articleDetail: ArticleProps,
  data: ArticleProps[]
}
