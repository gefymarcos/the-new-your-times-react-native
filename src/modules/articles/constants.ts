export const ARTICLES_REQUEST = 'ARTICLES_REQUEST'
export const ARTICLES_FULLFILLED = 'ARTICLES_FULLFILLED'
export const ARTICLES_REJECTED = 'ARTICLES_REJECTED'
export const ARTICLE_DETAIL_FULLFILLED = 'ARTICLE_DETAIL_FULLFILLED'
export const CLEAR_ARTICLE_DETAIL = 'CLEAR_ARTICLE_DETAIL'
