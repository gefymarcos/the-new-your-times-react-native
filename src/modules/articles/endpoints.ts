const routes = {
  articles: (section?: string) => `${section}.json`,
};

export default routes;
