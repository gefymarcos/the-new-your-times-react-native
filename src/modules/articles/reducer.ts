import * as constants from './constants';
import { ArticlesStoreProps, ArticleProps } from './types'

const articleInitialState: ArticleProps = {
  title: '',
  abstract: '',
  url: ''
}

const initialState: ArticlesStoreProps = {
  articlesRequesting: false,
  errored: false,
  errorMsg: '',
  articleDetail: articleInitialState,
  data: [],
};

const articlesReducer = (state = Object.assign({}, initialState), action: any) => {
  const { type, payload } = action;
  switch (type) {
    case constants.ARTICLES_REQUEST:
      return { ...state, articlesRequesting: true, data: [] };
    case constants.ARTICLES_FULLFILLED:
      return { ...Object.assign({}, initialState), data: payload };
    case constants.ARTICLES_REJECTED:

    return {
        ...Object.assign({}, initialState),
        errored: true,
        errorMsg: payload,
      };
    case constants.ARTICLE_SELECTED_REQUEST:
      return { ...Object.assign({}, initialState), articlesRequesting: true };
    case constants.ARTICLE_SELECTED_FULLFILLED:
      return { ...Object.assign({}, initialState), data: payload };
    case constants.ARTICLE_SELECTED_REJECTED:
      return {
        ...Object.assign({}, initialState),
        errored: true,
        errorMsg: payload,
      };
    case constants.CLEAN_ARTICLES_REQUEST:
      return { ...state, articlesRequesting: true };
    case constants.CLEAN_ARTICLES_FULLFILLED:
      return { ...Object.assign({}, initialState) };
    case constants.CLEAN_ARTICLES_REJECTED:
      return {
        ...Object.assign({}, initialState),
        errored: true,
        errorMsg: payload,
      };

    case constants.ARTICLE_DETAIL_FULLFILLED:
      return { ...Object.assign({}, state), articleDetail: payload };
    case constants.CLEAR_ARTICLE_DETAIL:
      return { ...state, articlesRequesting: true, articleDetail: articleInitialState };
    default:
      return state;
  }
};

export default articlesReducer;
