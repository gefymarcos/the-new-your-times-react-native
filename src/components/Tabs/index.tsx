import * as React from 'react'
import { StyleSheet } from 'react-native'
import { createMaterialBottomTabNavigator } from '@react-navigation/material-bottom-tabs'
import { NavigationContainer } from '@react-navigation/native'
import { useDispatch } from 'react-redux';
import { loadArticlesAction } from '../../modules/articles/actions';
import Science from '../../pages/Science'
import Technology from '../../pages/Technology'
import theme from '../../styles/theme'

type MaterialBottomTabParams = {
  TabScience: undefined
  TabTechnology: undefined
}

const MaterialBottomTabs = createMaterialBottomTabNavigator<MaterialBottomTabParams>()

const Tabs = () => {
  const dispatch = useDispatch()

  return (
    <NavigationContainer>
      <MaterialBottomTabs.Navigator barStyle={styles.tabBar}>
        <MaterialBottomTabs.Screen
          name="TabScience"
          component={Science}
          options={{
            tabBarLabel: 'Science',
            tabBarIcon: 'test-tube',
            tabBarColor: theme.colors.lightGray,
          }}
          listeners={{
            tabPress: () => {
              dispatch(loadArticlesAction('science'));
            }
          }}
          key='TabScience'
        />
        <MaterialBottomTabs.Screen
          name="TabTechnology"
          component={Technology}
          options={{
            tabBarLabel: 'Technology',
            tabBarIcon: 'tablet-dashboard',
            tabBarColor: theme.colors.lightGray
          }}
          listeners={{
            tabPress: () => {
              dispatch(loadArticlesAction('technology'));
            }
          }}
          key='TabTechnology'
        />
      </MaterialBottomTabs.Navigator>
    </NavigationContainer>
  )
}

const styles = StyleSheet.create({
  tabBar: {
    backgroundColor: theme.colors.black
  }
})

export default Tabs
