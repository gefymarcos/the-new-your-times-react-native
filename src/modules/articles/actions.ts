/* eslint-disable @typescript-eslint/no-explicit-any */
import { Dispatch } from 'react'
import { get } from '../../api'
import * as constants from './constants'
import { ArticleProps } from './types'
import endpoints from './endpoints'
import { cardProps } from '../../components/Card'

const articlesRequest = () => ({
  type: constants.ARTICLES_REQUEST
})

const articlesFullFilled = (articles: ArticleProps[]) => {
  return {
    type: constants.ARTICLES_FULLFILLED,
    payload: articles
  }
}

const articlesFullRejected = (err: string) => {
  return {
    type: constants.ARTICLES_REJECTED,
    payload: err
  }
}

const fetchArticles = async (section: string) => {
  return await get({ url: endpoints.articles(section) })
}

const loadArticlesAction = (section: string) => {
  return async (dispatch: any) => {
    dispatch(articlesRequest())
    try {
      const articles: any = await fetchArticles(section)
      dispatch(articlesFullFilled(articles?.data?.results))
    } catch (err) {
      console.log('err at load articles list', err)
      dispatch(articlesFullRejected(err))
    }
  }
}

const showArticleDetailAction = ({ title, abstract, url }: cardProps) => {
  return async (dispatch: Dispatch<any>) => {
    dispatch({
      type: constants.ARTICLE_DETAIL_FULLFILLED,
      payload: { title, abstract, url }
    })
  }
}

const clearArticleSelectedAction = () => {
  return async (dispatch: Dispatch<any>) => {
    dispatch({ type: constants.CLEAR_ARTICLE_DETAIL })
  }
}

export {
  loadArticlesAction,
  showArticleDetailAction,
  clearArticleSelectedAction
}
