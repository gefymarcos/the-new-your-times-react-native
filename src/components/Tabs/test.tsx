import React from 'react'
import { fireEvent, render } from '@testing-library/react-native'
import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import Tabs from '.'
import { ArticleProps, ArticlesStoreProps } from 'modules/articles/types'

jest.useFakeTimers()

describe('<Tabs />', () => {
  const articleInitialState: ArticleProps = {
    title: '',
    abstract: '',
    url: ''
  }
  const initialState: ArticlesStoreProps = {
    articlesRequesting: false,
    errored: false,
    errorMsg: '',
    articleDetail: articleInitialState,
    data: []
  }
  const middlewares = [thunk]
  const mockStore = configureStore(middlewares)
  let store

  it('should render correctly', () => {
    const { getAllByText } = render(
      <Provider store={mockStore({})}>
        <Tabs />
      </Provider>
    )

    const tab1 = getAllByText(/Science/i)
    const tab2 = getAllByText(/Technology/i)

    expect(tab1).toBeDefined()
    expect(tab2).toBeDefined()
  })

  it('should render correctly based on snapshot', () => {
    store = mockStore(initialState)
    const tree = renderer
      .create(
        <Provider store={store}>
          <Tabs />
        </Provider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should dispatched action with press science tab', () => {
    store = mockStore(initialState)

    const { getAllByText } = render(
      <Provider store={store}>
        <Tabs />
      </Provider>
    )

    const tab1 = getAllByText(/Science/i)
    fireEvent.press(tab1[0])
    expect(store.getActions()).toEqual([{ type: 'ARTICLES_REQUEST' }])
  })

  it('should dispatched action with press technology tab', () => {
    store = mockStore(initialState)

    const { getAllByText } = render(
      <Provider store={store}>
        <Tabs />
      </Provider>
    )

    const tab2 = getAllByText(/Science/i)
    fireEvent.press(tab2[0])
    expect(store.getActions()).toEqual([{ type: 'ARTICLES_REQUEST' }])
  })
})
