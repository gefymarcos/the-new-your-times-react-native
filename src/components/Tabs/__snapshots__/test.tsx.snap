// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> should render correctly based on snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "overflow": "hidden",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "flex": 1,
        },
        Object {
          "backgroundColor": "rgb(242, 242, 242)",
        },
      ]
    }
  >
    <View
      accessibilityElementsHidden={false}
      collapsable={false}
      importantForAccessibility="auto"
      pointerEvents="auto"
      removeClippedSubviews={false}
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "opacity": 1,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "top": 0,
          }
        }
      />
    </View>
  </View>
  <View
    onLayout={[Function]}
    pointerEvents="none"
    style={
      Object {
        "backgroundColor": "#212121",
        "bottom": 0,
        "elevation": 4,
        "left": 0,
        "position": null,
        "right": 0,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 3,
          "width": 0,
        },
        "shadowOpacity": 0.24,
        "shadowRadius": 4,
        "transform": Array [
          Object {
            "translateY": 0,
          },
        ],
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#212121",
          "overflow": "hidden",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
              "width": "100%",
            },
            Object {
              "marginBottom": 0,
              "maxWidth": 336,
            },
          ]
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "selected": true,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "flex": 1,
              "paddingVertical": 6,
            }
          }
        >
          <View
            pointerEvents="none"
          >
            <View
              style={
                Object {
                  "alignSelf": "center",
                  "height": 24,
                  "marginHorizontal": 12,
                  "marginTop": 2,
                  "transform": Array [
                    Object {
                      "translateY": 0,
                    },
                  ],
                  "width": 24,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "bottom": 0,
                    "left": 0,
                    "opacity": 1,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                      },
                      Object {
                        "color": "#ffffff",
                        "fontSize": 24,
                      },
                    ]
                  }
                >
                  □
                </Text>
              </View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "bottom": 0,
                    "left": 0,
                    "opacity": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                      },
                      Object {
                        "color": "rgba(255, 255, 255, 0.5)",
                        "fontSize": 24,
                      },
                    ]
                  }
                >
                  □
                </Text>
              </View>
              <View
                style={
                  Array [
                    Object {
                      "left": 0,
                      "position": "absolute",
                      "top": -2,
                    },
                    Object {
                      "right": -2,
                    },
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  style={
                    Object {
                      "alignSelf": "flex-end",
                      "backgroundColor": "#f50057",
                      "borderRadius": 8,
                      "color": "#ffffff",
                      "fontFamily": "System",
                      "fontSize": 8,
                      "fontWeight": "400",
                      "height": 16,
                      "lineHeight": 16,
                      "minWidth": 16,
                      "opacity": 0,
                      "overflow": "hidden",
                      "paddingHorizontal": 4,
                      "textAlign": "center",
                      "textAlignVertical": "center",
                    }
                  }
                />
              </View>
            </View>
            <View
              style={
                Object {
                  "height": 16,
                  "paddingBottom": 2,
                  "transform": Array [
                    Object {
                      "scale": 1,
                    },
                  ],
                }
              }
            >
              <View
                style={
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "opacity": 1,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontWeight": "400",
                      },
                      Object {
                        "textAlign": "left",
                      },
                      Array [
                        Object {
                          "backgroundColor": "transparent",
                          "fontSize": 12,
                          "textAlign": "center",
                        },
                        Object {
                          "color": "#ffffff",
                        },
                      ],
                    ]
                  }
                >
                  Science
                </Text>
              </View>
              <View
                style={
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "opacity": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
              >
                <Text
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontWeight": "400",
                      },
                      Object {
                        "textAlign": "left",
                      },
                      Array [
                        Object {
                          "backgroundColor": "transparent",
                          "fontSize": 12,
                          "textAlign": "center",
                        },
                        Object {
                          "color": "rgba(255, 255, 255, 0.5)",
                        },
                      ],
                    ]
                  }
                >
                  Science
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "selected": true,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "flex": 1,
              "paddingVertical": 6,
            }
          }
        >
          <View
            pointerEvents="none"
          >
            <View
              style={
                Object {
                  "alignSelf": "center",
                  "height": 24,
                  "marginHorizontal": 12,
                  "marginTop": 2,
                  "transform": Array [
                    Object {
                      "translateY": 0,
                    },
                  ],
                  "width": 24,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "bottom": 0,
                    "left": 0,
                    "opacity": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                      },
                      Object {
                        "color": "#ffffff",
                        "fontSize": 24,
                      },
                    ]
                  }
                >
                  □
                </Text>
              </View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "bottom": 0,
                    "left": 0,
                    "opacity": 1,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                      },
                      Object {
                        "color": "rgba(255, 255, 255, 0.5)",
                        "fontSize": 24,
                      },
                    ]
                  }
                >
                  □
                </Text>
              </View>
              <View
                style={
                  Array [
                    Object {
                      "left": 0,
                      "position": "absolute",
                      "top": -2,
                    },
                    Object {
                      "right": -2,
                    },
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  style={
                    Object {
                      "alignSelf": "flex-end",
                      "backgroundColor": "#f50057",
                      "borderRadius": 8,
                      "color": "#ffffff",
                      "fontFamily": "System",
                      "fontSize": 8,
                      "fontWeight": "400",
                      "height": 16,
                      "lineHeight": 16,
                      "minWidth": 16,
                      "opacity": 0,
                      "overflow": "hidden",
                      "paddingHorizontal": 4,
                      "textAlign": "center",
                      "textAlignVertical": "center",
                    }
                  }
                />
              </View>
            </View>
            <View
              style={
                Object {
                  "height": 16,
                  "paddingBottom": 2,
                  "transform": Array [
                    Object {
                      "scale": 1,
                    },
                  ],
                }
              }
            >
              <View
                style={
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "opacity": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontWeight": "400",
                      },
                      Object {
                        "textAlign": "left",
                      },
                      Array [
                        Object {
                          "backgroundColor": "transparent",
                          "fontSize": 12,
                          "textAlign": "center",
                        },
                        Object {
                          "color": "#ffffff",
                        },
                      ],
                    ]
                  }
                >
                  Technology
                </Text>
              </View>
              <View
                style={
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "opacity": 1,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
              >
                <Text
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontWeight": "400",
                      },
                      Object {
                        "textAlign": "left",
                      },
                      Array [
                        Object {
                          "backgroundColor": "transparent",
                          "fontSize": 12,
                          "textAlign": "center",
                        },
                        Object {
                          "color": "rgba(255, 255, 255, 0.5)",
                        },
                      ],
                    ]
                  }
                >
                  Technology
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`
